src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/DiagnosticUtilities.cs

```diff
@@ -44,40 +50,110 @@ internal static string GetGenericParameterDeclaringMemberDisplayName(GenericPara
         {
-            var ecmaMethod = method.GetTypicalMethodDefinition() as EcmaMethod;
-            if (ecmaMethod == null)
-                return null;
+            attribute = default;
+            CustomAttributeValue<TypeDesc>? decoded = default;
+            switch (member)
+            {
+                case MethodDesc method:
+                    var ecmaMethod = method.GetTypicalMethodDefinition() as EcmaMethod;
+                    if (ecmaMethod == null)
+                        return false;
+                    decoded = ecmaMethod.GetDecodedCustomAttribute("System.Diagnostics.CodeAnalysis", requiresAttributeName);
+                    break;
+                case MetadataType type:
+                    var ecmaType = type as EcmaType;
+                    if (ecmaType == null)
+                        return false;
+                    decoded = ecmaType.GetDecodedCustomAttribute("System.Diagnostics.CodeAnalysis", requiresAttributeName);
+                    break;
+                default:
+                    Debug.Fail(member.GetType().ToString());
+                    break;
+            }
+            if (!decoded.HasValue)
+                return false;
+
+            attribute = decoded.Value;
+            return true;
+        }
 
-            var decoded = ecmaMethod.GetDecodedCustomAttribute("System.Diagnostics.CodeAnalysis", requiresAttributeName);
-            if (decoded == null)
-                return null;
+        internal static string GetRequiresAttributeMessage(CustomAttributeValue<TypeDesc> attribute)
+        {
+            if (attribute.FixedArguments.Length != 0)
+                return (string)attribute.FixedArguments[0].Value;
 
-            var decodedValue = decoded.Value;
+            return null;
+        }
 
-            if (decodedValue.FixedArguments.Length != 0)
-                return (string)decodedValue.FixedArguments[0].Value;
+        internal static string GetRequiresAttributeUrl(CustomAttributeValue<TypeDesc> attribute)
+        {
+            if (attribute.NamedArguments.Length != 0 && attribute.NamedArguments[0].Name == "Url")
+                return (string)attribute.NamedArguments[0].Value;
 
             return null;
         }
 
-        internal static string GetRequiresAttributeUrl(MethodDesc method, string requiresAttributeName)
+        /// <summary>
+        /// Determines if method is within a declared Requires scope - this typically means that trim analysis
+        /// warnings should be suppressed in such a method.
+        /// </summary>
+        /// <remarks>Unlike <see cref="DoesMethodRequires(MethodDesc, string, out CustomAttributeValue?)"/>
+        /// if a declaring type has Requires, all methods in that type are considered "in scope" of that Requires. So this includes also
+        /// instance methods (not just statics and .ctors).</remarks>
+        internal static bool IsMethodInRequiresScope(this MethodDesc method, string requiresAttribute)
```

Comment:
I would call this IsInRequiresScope - that's the name we use in the analyzer - so it will be easier to lookup the matching functionality there. (The body does match the behavior, so the name might as well)

Reply:
I see the linker calls this `IsMethodInRequiresUnreferencedCodeScope`...
In that case I don't mind the current name, but we should then rename the one in the analyzer to match.
And please add a comment which methods this is a counterpart for.

Reply:
Opened https://github.com/dotnet/linker/pull/2788 to track the changes between linker/analyzer